File: 0-Strings/476-Number-Complement/476-Number-Complement.java
class Solution {
    public int findComplement(int num) {
        int i = 0;
        int j = 0;
        while (i < num)
        {
            i += Math.pow(2, j);
            j++;
        }
        return i - num;
    }
}

/**
 * Any binary representation of a number, its complement
 * is just the smallest number n of the form 11..11
 * minus that number. 
 * E.g. 94 = 1011110 -> 1111111 - 1011110 = 100001 = 33
 */

File: 0-Strings/476-Number-Complement/476-Number-Complement.cpp
/**
 * Problem 476: Number Complement
 * Prompt: Given a positive integer, output its complement number.
 * The complement strategy is to flip the bits of its binary representation.
 * Date: 05/04/2020
 * */

#include <sstream>
#include <string>

using std::stringstream;

class Solution {
public:
    int findComplement(int num) {
        stringstream ss;
        while (num != 0)
        {
            int dig = num % 2;
            if (dig)
                ss << 0;
            else
                ss << 1;
            num = num / 2;            
        }
        string ans = ss.str();
        reverse(ans.begin(), ans.end());
        return stoi(ans,0,2);
    }
};

/**
 * Note: Since the algorithm of finding the binary representation
 * geneates the reverse of binaray representation, there needs
 * to be a reverse
 * Then use the function to convert it back to integer
 * */